internal/internal_update.go

```diff
@@ -207,37 +217,47 @@ func defaultUpdateHandler(
 	serializedArgs *commonpb.Payloads,
 	header *commonpb.Header,
 	callbacks UpdateCallbacks,
-	spawn func(Context, string, func(Context)) Context,
+	scheduler UpdateScheduler,
 ) {
 	env := getWorkflowEnvironment(rootCtx)
 	ctx, err := workflowContextWithHeaderPropagated(rootCtx, header, env.GetContextPropagators())
 	if err != nil {
 		callbacks.Reject(err)
 		return
 	}
-	eo := getWorkflowEnvOptions(ctx)
-	handler, ok := eo.updateHandlers[name]
-	if !ok {
-		keys := make([]string, 0, len(eo.updateHandlers))
-		for k := range eo.updateHandlers {
-			keys = append(keys, k)
+	scheduler.Spawn(ctx, name, func(ctx Context) {
+		eo := getWorkflowEnvOptions(ctx)
+
+		// If we suspect that handler registration has not occurred (e.g.
+		// because this update is part of the first workflow task and is being
+		// delivered before the workflow function itself has run and had a
+		// chance to register update handlers) then we yield control back to the
+		// scheduler to allow handler registration to occur. The scheduler will
+		// resume this coroutine after others have run to a blocking point.
+		if len(eo.updateHandlers) == 0 {
```

Comment:
I have a few of concerns here:

1. What are the delivery ordering semantics after adding this yield?
What if we have a signal-update-signal sequence?
With this solution, will the update necessarily be delivered after the signals?

2. What about the future async update guarantees?
Are we okay rejecting any updates if a handler is not registered by the end of a workflow task? This is different than signals, which are unidirectional and are okay to buffer.

3. What if user wants to run a local activity before registering a dynamic update handler?
Does using this one time yield mechanism prevent that case?

I'm leaning towards having a solution where we buffer updates and (at least with the fast path) we reject them at the end of workflow task processing.